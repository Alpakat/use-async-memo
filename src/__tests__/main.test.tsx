import * as testing from '@testing-library/react'
import {FC, useState} from 'react'
import * as React from 'react'
import {useAsyncMemo} from '../index'
import {act} from '@testing-library/react'
import {useDebounce} from 'use-debounce'

test('get async data', async function () {
  function getAsyncData() {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve(1)
      }, 100)
    })
  }

  const App: FC = function (props) {
    const data = useAsyncMemo(() => getAsyncData(), [])
    return (
      <div>
        {data}
      </div>
    )
  }
  const renderer = testing.render(
    <App/>
  )
  expect(renderer.asFragment()).toMatchSnapshot()
  await getAsyncData()
  expect(renderer.asFragment()).toMatchSnapshot()
})


test('with debounce', async function () {
  function getAsyncData(x: number) {
    return new Promise<number>((resolve) => {
      setTimeout(() => {
        resolve(x)
      }, 50)
    })
  }

  let change: () => void

  const App: FC = function (props) {
    const [x, setX] = useState(1)
    const [debouncedX] = useDebounce(x, 100)

    const y = useAsyncMemo(async () => {
      return await getAsyncData(debouncedX)
    }, [debouncedX])

    change = () => {
      setX(x + 1)
    }

    return (
      <div>
        <span>{x}</span>
        <span>{y}</span>
      </div>
    )
  }
  const renderer = testing.render(
    <App/>
  )
  expect(renderer.asFragment()).toMatchSnapshot()
  act(() => {
    change()
  })
  expect(renderer.asFragment()).toMatchSnapshot()
  change()
  change()
  change()
  await new Promise((resolve) => {
    setTimeout(resolve, 1000)
  })
  expect(renderer.asFragment()).toMatchSnapshot()
})

