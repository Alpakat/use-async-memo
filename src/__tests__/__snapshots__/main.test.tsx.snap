// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get async data 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`get async data 2`] = `
<DocumentFragment>
  <div>
    1
  </div>
</DocumentFragment>
`;

exports[`with debounce 1`] = `
<DocumentFragment>
  <div>
    <span>
      1
    </span>
    <span />
  </div>
</DocumentFragment>
`;

exports[`with debounce 2`] = `
<DocumentFragment>
  <div>
    <span>
      2
    </span>
    <span />
  </div>
</DocumentFragment>
`;

exports[`with debounce 3`] = `
<DocumentFragment>
  <div>
    <span>
      5
    </span>
    <span>
      5
    </span>
  </div>
</DocumentFragment>
`;
